packages/backend-common/src/database/postgres.ts

```diff
@@ -15,26 +15,41 @@
  */
 
 import knex, { PgConnectionConfig } from 'knex';
-import { Config } from '@backstage/config';
+import { cloneDeep } from 'lodash';
+import { Config, JsonValue } from '@backstage/config';
 import { mergeDatabaseConfig } from './config';
 
 /**
- * Creates a knex postgres database connection
+ * Creates a knex Postgres database connection
  *
  * @param dbConfig The database config
  * @param overrides Additional options to merge with the config
  */
-export function createPgDatabaseClient(
+export async function createPgDatabaseClient(
   dbConfig: Config,
   overrides?: knex.Config,
 ) {
-  const knexConfig = buildPgDatabaseConfig(dbConfig, overrides);
+  const baseConfig = buildPgDatabaseConfig(dbConfig, overrides);
+  let knexConfig = baseConfig;
+
+  // Bootstrap the missing database.
+  if (!!baseConfig?.connection.database) {
```

Comment:
I'm wondering if assuming no database is the best approach here. When I thought about this, I was thinking about making this an explicitly optional step that would be done in the `packages/backend/src/index.ts` file.  To me, doing this here makes the connection code more complex and tightly couples the concept of creating a connection and roles.

I'm wondering if it might be better to invert this so the admin role is a different module that uses this connection logic rather then tying the 2 together. Then maybe the `ensureDb` related functions could also be another module and used explicitly in the backend as a first step. This would also allow the org instance to compose where the connection comes from for admin vs app more clearly.

Reply:
So as a really general concept something like:

```ts
// somewhere in packages/backend/src/index.ts
const adminClient = createDatabaseAdminClient(config.getConfig('backend.adminDatabase'))
const dbName = `backstage_plugin_${plugin}`;
await ensurePgDatabase(adminClient, dbName);
```

Reply:
I guess you only mean admin credentials here? To me the `adminDatabase` sounds like a separate database with it's own host and port config.

Reply:
It could be the `postgres` database for instance. It's really just a knex connection to some database on the server that has admin like permissions.

Reply:
so @mfrinnstrom you mean database as in "physical" database, and @andrewthauer you mean as in "logical" database (as in CREATE DATABASE) within the same server right? Am I understanding that right

Reply:
@freben - Correct (within the same server). I'm not sure we are trying to tackle a different server per plugin database here are we? However, that is potentially an interesting point. But to my approach here, keeping the connection helpers very single purpose could aid in this as well (if we choose to support it).

Reply:
@freben @mfrinnstrom @andrewthauer to disambiguate: can we standardise on:
 - database *server*: the RDS/postgres service itself
 - database: the `CREATE DATABASE X` bit

I think we would need to normalise the terminology in several places.